### DISCUSSION TITLE
Good testing setup with Mocha?

### DISCUSSION THREAD

**Discussion by Nikola-Milovic** [typegoose/mongodb-memory-server#584] [Q&A] [answered]
Hello everyone, what is a good approach when it comes to testing with Mocha and MongoDB? Mongodb memory server seems like the perfect fit, but I've been unable to get it setup properly. 

Do you setup your server once before all tests or before each test/ test block? As Unit tests and integration tests for that manner, should have a clean DB to avoid unusual behaviour. 

This is 50% Mocha and 50% MDB Memory server question but I am guessing I'll have more luck asking here. If anyone would be willing to share their setup, mine can be [found here alongside the issues I am facing.](https://stackoverflow.com/q/70257212/13440079)

**Comment by hasezoey** [comment, marked as answer]
i dont quite know you specifications, but there is [Integration with Test Runners: Mocha / chai](https://nodkz.github.io/mongodb-memory-server/docs/guides/integration-examples/test-runners#mocha--chai), which may be outdated, because we (maintainers) dont really use mocha / chai anymore, but there is a [Integration with Test Runners: jest](https://nodkz.github.io/mongodb-memory-server/docs/guides/integration-examples/test-runners#jest) which is up-to-date

> Do you setup your server once before all tests or before each test/ test block?

this depends on what you want from your tests:
- Do you want Fast tests (and dont care about carrying data over to other tests / test suites)? then, a global instance should be used
- Do you want Accurate tests (and care about carrying data over to other tests / test suites)? then a instance for each test suite (or test) should be used (but is way slower)

i personally recommend to use a global instance if possible, because it is way faster for the test than spinning up and cleaning up instance after each test / test suite and also currently it may fail (randomly?) if tests are run parallel because ports may become used after being checked for being free

also, there is the possibility of just using different databases for each test suite / test and/or running a `database.delete` (or whatever it is called) after each / before each test / test suite
